crypto/sm4/sm4.c

```diff
@@ -0,0 +1,232 @@
+   0x3C0C3030, 0xCE945A5A, 0xAB76DDDD, 0x867FF9F9, 0xF1649595, 0x5DBBE6E6,
+   0x35F2C7C7, 0x2D092424, 0xD1C61717, 0xD66FB9B9, 0xDEC51B1B, 0x94861212,
+   0x78186060, 0x30F3C3C3, 0x897CF5F5, 0x5CEFB3B3, 0xD23AE8E8, 0xACDF7373,
+   0x794C3535, 0xA0208080, 0x9D78E5E5, 0x56EDBBBB, 0x235E7D7D, 0xC63EF8F8,
+   0x8BD45F5F, 0xE7C82F2F, 0xDD39E4E4, 0x68492121 };
+
+static inline uint32_t rotl(uint32_t a, uint8_t n)
+{
+    return (a << n) | (a >> (32 - n));
+}
+
+static inline uint32_t load_u32_be(const uint8_t *b, uint32_t n)
+{
+    return ((uint32_t)b[4 * n] << 24) ^
+           ((uint32_t)b[4 * n + 1] << 16) ^
+           ((uint32_t)b[4 * n + 2] << 8) ^
+           ((uint32_t)b[4 * n + 3]);
+}
+
+static inline void store_u32_be(uint32_t v, uint8_t *b)
+{
+    b[0] = (v >> 24) & 0xFF;
+    b[1] = (v >> 16) & 0xFF;
+    b[2] = (v >> 8) & 0xFF;
+    b[3] = (v) & 0xFF;
+}
+
+static inline uint32_t SM4_T_slow(uint32_t X)
+{
+    uint32_t t = 0;
+
+    t |= ((uint32_t)SM4_S[(X >> 24) & 0xFF]) << 24;
+    t |= ((uint32_t)SM4_S[(X >> 16) & 0xFF]) << 16;
+    t |= ((uint32_t)SM4_S[(X >> 8) & 0xFF]) << 8;
+    t |= SM4_S[X & 0xFF];
+
+    /*
+     * L linear transform
+     */
+    return t ^ rotl(t, 2) ^ rotl(t, 10) ^ rotl(t, 18) ^ rotl(t, 24);
+}
+
+static inline uint32_t SM4_T(uint32_t X)
+{
+    return SM4_SBOX_T[(X >> 24) & 0xFF] ^
+           rotl(SM4_SBOX_T[(X >> 16) & 0xFF], 24) ^
+           rotl(SM4_SBOX_T[(X >> 8) & 0xFF], 16) ^
+           rotl(SM4_SBOX_T[X & 0xFF], 8);
+}
+
+int SM4_set_key(const uint8_t *key, SM4_KEY *ks)
+{
+    /*
+     * Family Key
+     */
+    static const uint32_t FK[4] =
+        { 0xa3b1bac6, 0x56aa3350, 0x677d9197, 0xb27022dc };
+
+    /*
+     * Constant Key
+     */
+    static const uint32_t CK[32] = {
+        0x00070E15, 0x1C232A31, 0x383F464D, 0x545B6269,
+        0x70777E85, 0x8C939AA1, 0xA8AFB6BD, 0xC4CBD2D9,
+        0xE0E7EEF5, 0xFC030A11, 0x181F262D, 0x343B4249,
+        0x50575E65, 0x6C737A81, 0x888F969D, 0xA4ABB2B9,
+        0xC0C7CED5, 0xDCE3EAF1, 0xF8FF060D, 0x141B2229,
+        0x30373E45, 0x4C535A61, 0x686F767D, 0x848B9299,
+        0xA0A7AEB5, 0xBCC3CAD1, 0xD8DFE6ED, 0xF4FB0209,
+        0x10171E25, 0x2C333A41, 0x484F565D, 0x646B7279
+    };
+
+    uint32_t K[4];
+    uint8_t i;
+
+    K[0] = load_u32_be(key, 0) ^ FK[0];
+    K[1] = load_u32_be(key, 1) ^ FK[1];
+    K[2] = load_u32_be(key, 2) ^ FK[2];
+    K[3] = load_u32_be(key, 3) ^ FK[3];
+
+    for (i = 0; i != SM4_KEY_SCHEDULE; ++i) {
+        uint32_t X = K[(i + 1) % 4] ^ K[(i + 2) % 4] ^ K[(i + 3) % 4] ^ CK[i];
+        uint32_t t = 0;
+
+        t |= ((uint32_t)SM4_S[(X >> 24) & 0xFF]) << 24;
+        t |= ((uint32_t)SM4_S[(X >> 16) & 0xFF]) << 16;
+        t |= ((uint32_t)SM4_S[(X >> 8) & 0xFF]) << 8;
+        t |= SM4_S[X & 0xFF];
+
+        t = t ^ rotl(t, 13) ^ rotl(t, 23);
+        K[i % 4] ^= t;
+        ks->rk[i] = K[i % 4];
+    }
+
+    return 1;
+}
+
+#define SM4_RNDS(k0, k1, k2, k3, F) do {     \
```

Comment:
I guess it's preferred if the `do` statement is placed in next line as:
```
    #define macrofun(a, b, c)   \
        do {                    \
            if (a == 5)         \
                do_this(b, c);  \
        } while (0)
```

Reply:
Addressed.